chameleon-core/src/main/java/org/hibernate/omm/ast/mql/AttachmentKeys.java

```diff
@@ -0,0 +1,103 @@
+package org.hibernate.omm.ast.mql;
+
+import org.hibernate.omm.mongoast.AstSortField;
+import org.hibernate.omm.mongoast.AstValue;
+import org.hibernate.omm.mongoast.filters.AstFilter;
+import org.hibernate.omm.mongoast.stages.AstLookupStage;
+import org.hibernate.omm.mongoast.stages.AstProjectStageSpecification;
+import org.hibernate.omm.mongoast.stages.AstStage;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static com.mongodb.assertions.Assertions.assertTrue;
+
+public class AttachmentKeys {
+
+    private static final AttachmentKey<String> COLLECTION_NAME = new DefaultAttachmentKey<>("collectionName");
+    private static final AttachmentKey<List<AstProjectStageSpecification>> PROJECT_STAGE_SPECIFICATIONS =
+            new DefaultAttachmentKey<>("projectStageSpecifications");
+    private static final AttachmentKey<String> FIELD_NAME = new DefaultAttachmentKey<>("fieldName");
+    private static final AttachmentKey<AstValue> FIELD_VALUE = new DefaultAttachmentKey<>("fieldValue");
+    private static final AttachmentKey<AstFilter> FILTER = new DefaultAttachmentKey<>("filter");
+    private static final AttachmentKey<List<AstSortField>> SORT_FIELDS = new DefaultAttachmentKey<>("sortFields");
+    private static final AttachmentKey<AstSortField> SORT_FIELD = new DefaultAttachmentKey<>("sortField");
+    private static final AttachmentKey<List<AstStage>> JOIN_STAGES = new DefaultAttachmentKey<>("joinStages");
+    private static final AttachmentKey<AstLookupStage> LOOKUP_STAGE = new DefaultAttachmentKey<>("lookupStage");
+    private static final AttachmentKey<CollectionNameAndJoinStages> COLLECTION_NAME_AND_JOIN_STAGES = new DefaultAttachmentKey<>("collectionNameAndJoinStages");
+
```

Comment:
so from my understanding, the concerns here are twofold:

- generic type
- name

If we only care about generic type, the `TypeReference` class from Jackson library suffices (e.g. see its usage in Jackson's deserialization: https://www.baeldung.com/java-deserialize-generic-type-with-jackson). The Attachment related classes are a little bit convoluted to my taste. I would simplify it as much as possible to the following one single Attachment class:
```
public class Attachment {
    public record Expectation<T> (TypeReference<T> type, String label){};
   
    private Expectation<?> expectation;
    private Object curValue;
 
    public <T> void attach(Expectation<T> expectation, T value) {
        ... ...
    }
    public <T> T expect(Expectation<T> expectation, Runnable runnable) {
        ... ...
    }
}
```
but still it is anything but elegant. Seems there is no silver bullet.